File: libandroid-navigation-ui/src/test/java/org/maplibre/navigation/android/navigation/ui/v5/TestRouteProgressBuilder.java
package org.maplibre.navigation.android.navigation.ui.v5;

import static org.maplibre.navigation.android.navigation.ui.v5.GeoJsonExtKt.toJvmPoints;
import static org.maplibre.navigation.core.navigation.NavigationHelper.createDistancesToIntersections;
import static org.maplibre.navigation.core.navigation.NavigationHelper.createIntersectionsList;
import static org.maplibre.navigation.core.utils.Constants.PRECISION_6;

import androidx.annotation.NonNull;

import org.maplibre.navigation.core.models.DirectionsRoute;
import org.maplibre.navigation.core.models.LegStep;
import org.maplibre.navigation.core.models.StepIntersection;
import org.maplibre.geojson.Point;
import org.maplibre.geojson.utils.PolylineUtils;
import org.maplibre.navigation.core.routeprogress.RouteProgress;

import java.util.List;
import java.util.Map;

class TestRouteProgressBuilder {

    RouteProgress buildDefaultTestRouteProgress(DirectionsRoute testRoute) {
        return buildTestRouteProgress(testRoute, 100, 100,
            100, 0, 0);
    }

    RouteProgress buildTestRouteProgress(DirectionsRoute route,
                                         double stepDistanceRemaining,
                                         double legDistanceRemaining,
                                         double distanceRemaining,
                                         int stepIndex,
                                         int legIndex) {
        List<LegStep> steps = route.getLegs().get(legIndex).getSteps();
        LegStep currentStep = steps.get(stepIndex);
        List<Point> currentStepPoints = buildCurrentStepPoints(currentStep);
        int upcomingStepIndex = stepIndex + 1;
        List<Point> upcomingStepPoints = null;
        LegStep upcomingStep = null;
        if (upcomingStepIndex < steps.size()) {
            upcomingStep = steps.get(upcomingStepIndex);
            String upcomingStepGeometry = upcomingStep.getGeometry();
            upcomingStepPoints = buildStepPointsFromGeometry(upcomingStepGeometry);
        }

        List<StepIntersection> intersections = createIntersectionsList(currentStep, upcomingStep);
        Map<StepIntersection, Double> intersectionDistances = createDistancesToIntersections(
            toJvmPoints(currentStepPoints), intersections
        );

        return new RouteProgress.Builder(
            route,
            legIndex,
            distanceRemaining,
            toJvmPoints(currentStepPoints),
            stepIndex,
            legDistanceRemaining,
            stepDistanceRemaining
        )
            .withUpcomingStepPoints(upcomingStepPoints)
            .withIntersections(intersections)
            .withCurrentIntersection(intersections.get(0))
            .withIntersectionDistancesAlongStep(intersectionDistances)
            .build();
    }

    @NonNull
    private List<Point> buildCurrentStepPoints(LegStep currentStep) {
        String currentStepGeometry = currentStep.getGeometry();
        return buildStepPointsFromGeometry(currentStepGeometry);
    }

    private List<Point> buildStepPointsFromGeometry(String stepGeometry) {
        return toJvmPoints(PolylineUtils.decode(stepGeometry, PRECISION_6));
    }
}

File: libandroid-navigation/src/test/java/org/maplibre/navigation/android/navigation/v5/TestRouteProgressBuilder.kt
package org.maplibre.navigation.android.navigation.v5

import org.maplibre.geojson.Point
import org.maplibre.geojson.utils.PolylineUtils
import org.maplibre.navigation.android.navigation.v5.models.DirectionsRoute
import org.maplibre.navigation.android.navigation.v5.models.LegStep
import org.maplibre.navigation.android.navigation.v5.models.StepIntersection
import org.maplibre.navigation.android.navigation.v5.navigation.NavigationHelper.createDistancesToIntersections
import org.maplibre.navigation.android.navigation.v5.navigation.NavigationHelper.createIntersectionsList
import org.maplibre.navigation.android.navigation.v5.navigation.NavigationHelper.findCurrentIntersection
import org.maplibre.navigation.android.navigation.v5.navigation.NavigationHelper.findUpcomingIntersection
import org.maplibre.navigation.android.navigation.v5.routeprogress.RouteProgress
import org.maplibre.navigation.android.navigation.v5.utils.Constants

internal class TestRouteProgressBuilder {
    @Throws(Exception::class)
    fun buildDefaultTestRouteProgress(testRoute: DirectionsRoute): RouteProgress {
        return buildTestRouteProgress(
            testRoute, 100.0, 100.0,
            100.0, 0, 0
        )
    }

    @Throws(Exception::class)
    fun buildTestRouteProgress(
        route: DirectionsRoute,
        stepDistanceRemaining: Double,
        legDistanceRemaining: Double,
        distanceRemaining: Double,
        stepIndex: Int,
        legIndex: Int
    ): RouteProgress {
        val steps = route.legs[legIndex].steps
        val currentStep = steps[stepIndex]
        val currentStepPoints = buildCurrentStepPoints(currentStep)
        val upcomingStepIndex = stepIndex + 1
        var upcomingStepPoints: List<Point>? = null
        var upcomingStep: LegStep? = null
        if (upcomingStepIndex < steps.size) {
            upcomingStep = steps[upcomingStepIndex]
            val upcomingStepGeometry = upcomingStep.geometry
            upcomingStepPoints = buildStepPointsFromGeometry(upcomingStepGeometry)
        }
        val intersections: List<StepIntersection> =
            createIntersectionsList(currentStep, upcomingStep)
        val intersectionDistances = createDistancesToIntersections(
            currentStepPoints, intersections
        )

        val currentIntersection = createCurrentIntersection(
            stepDistanceRemaining,
            currentStep,
            intersections,
            intersectionDistances
        )
        val upcomingIntersection = createUpcomingIntersection(
            upcomingStep,
            intersections,
            currentIntersection!!
        )

        return RouteProgress(
            stepDistanceRemaining = stepDistanceRemaining,
            legDistanceRemaining = legDistanceRemaining,
            distanceRemaining = distanceRemaining,
            directionsRoute = route,
            currentStepPoints = currentStepPoints,
            upcomingStepPoints = upcomingStepPoints,
            intersections = intersections,
            currentIntersection = currentIntersection,
            upcomingIntersection = upcomingIntersection,
            intersectionDistancesAlongStep = intersectionDistances,
            stepIndex = stepIndex,
            legIndex = legIndex,
            currentLegAnnotation = null
        )
    }

    private fun buildCurrentStepPoints(currentStep: LegStep): List<Point> {
        val currentStepGeometry = currentStep.geometry
        return buildStepPointsFromGeometry(currentStepGeometry)
    }

    private fun createCurrentIntersection(
        stepDistanceRemaining: Double, currentStep: LegStep,
        intersections: List<StepIntersection>,
        intersectionDistances: Map<StepIntersection, Double>
    ): StepIntersection? {
        val stepDistanceTraveled = currentStep.distance - stepDistanceRemaining
        return findCurrentIntersection(
            intersections,
            intersectionDistances, stepDistanceTraveled
        )
    }

    private fun createUpcomingIntersection(
        upcomingStep: LegStep?,
        intersections: List<StepIntersection>,
        currentIntersection: StepIntersection
    ): StepIntersection? {
        return findUpcomingIntersection(
            intersections,
            upcomingStep,
            currentIntersection
        )
    }

    private fun buildStepPointsFromGeometry(stepGeometry: String): List<Point> {
        return PolylineUtils.decode(stepGeometry, Constants.PRECISION_6)
    }
}
